datafusion/physical-optimizer/src/progressive_evaluation/lexical_ranges.rs

```diff
@@ -0,0 +1,774 @@
+}
+
+impl LexicalRange {
+    pub fn builder() -> LexicalRangeBuilder {
+        LexicalRangeBuilder::new()
+    }
+}
+
+impl Display for LexicalRange {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "({})->({})",
+            self.min
+                .iter()
+                .map(|v| v.to_string())
+                .collect::<Vec<_>>()
+                .join(","),
+            self.max
+                .iter()
+                .map(|v| v.to_string())
+                .collect::<Vec<_>>()
+                .join(",")
+        )
+    }
+}
+
+/// Builder for [`LexicalRange`]
+#[derive(Debug, Default, Clone)]
+pub struct LexicalRangeBuilder {
+    inner: LexicalRange,
+}
+
+impl LexicalRangeBuilder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn build(self) -> LexicalRange {
+        self.inner
+    }
+
+    /// Adds min and max to the end of the in-progress ranges
+    pub fn push(&mut self, min: ScalarValue, max: ScalarValue) {
+        self.inner.min.push(min);
+        self.inner.max.push(max);
+    }
+}
+
+/// Represents ranges of lexically ordered values in a sort order.
+///
+/// One element for each input partition
+
+#[derive(Debug)]
+pub struct NonOverlappingOrderedLexicalRanges {
+    /// Corresponding lexical range per partition, already reordered to match indices
+    value_ranges: Vec<LexicalRange>,
+
+    /// The order of the input partitions (rows of ranges) that would provide ensure they are
+    /// sorted in lexical order
+    indices: Vec<usize>,
+}
+
+impl Display for NonOverlappingOrderedLexicalRanges {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "LexicalRanges\n  {:?}", self.indices,)
+    }
+}
+
+impl NonOverlappingOrderedLexicalRanges {
+    /// Attempt to create a new [`NonOverlappingOrderedLexicalRanges`]
+    ///
+    /// Returns None if:
+    /// * - There are no ranges
+    /// * - The ranges are not disjoint (aka they overlap in the lexical space)
+    ///
+    /// Returns Err if there is an error converting values
+    pub fn try_new(
+        sort_options: &[SortOptions],
+        ranges_per_partition: Vec<LexicalRange>,
+    ) -> Result<Option<Self>> {
+        if ranges_per_partition.is_empty() {
+            return Ok(None);
+        }
+        // convert to min/maxes, as VecPerPartition<VecPerSortKey>
+        let (mins, maxes) = ranges_per_partition.clone().into_iter().fold(
+            (vec![], vec![]),
+            |(mut mins, mut maxes), partition_range| {
+                let LexicalRange {
+                    min: min_per_sort_key,
+                    max: max_per_sort_key,
+                } = partition_range;
+                mins.push(min_per_sort_key);
+                maxes.push(max_per_sort_key);
+                (mins, maxes)
+            },
+        );
+        let rows = ConvertedRows::try_new(sort_options, mins, maxes)?;
```

Comment:
This `ConvertedRows` is where we apply a sense of ordering using the same RowConverter (to compare [Rows](https://docs.rs/datafusion/latest/datafusion/common/arrow/row/struct.Row.html)).